src/utils/date.ts

```diff
@@ -0,0 +1,74 @@
+import { isString } from "@chakra-ui/utils";
+import { RangeValue } from "@react-types/shared";
+
+// As per https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#dates
+export function parseDate(dateValue: string | undefined) {
+  if (dateValue == null || !isString(dateValue)) return;
+
+  const dateRegex = dateValue.match(/(\d{4,})(-(\d\d)){1,2}/i);
+  if (dateRegex == null) return;
+
+  const date = dateValue.split("-");
+  if (+date[0] <= 0 || +date[1] < 1 || +date[1] > 12) return;
+
+  const maxDay = isLeapYear(+date[0])
+    ? getMaxDay(+date[1], true)
+    : getMaxDay(+date[1]);
+  if (+date[2] < 1 && +date[2] > maxDay) return;
+
+  return new Date(
+    +date[0],
```

Comment:
We can probably use

```ts
export function parseDate(dateValue: string | undefined) {
  if (!dateValue) return;
  return parseISO(dateValue as string);
}
```

Reply:
`parseISO` doesn't accept a value like this 

```html
<input type="date" id="start" name="trip-start" value="201800-07-22">
```

Which is valid according to https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#dates

![image](https://user-images.githubusercontent.com/39694575/95864798-8816bb00-0d83-11eb-8741-554c6cb08935.png)

So I used this instead,

```js
export function parseDate(dateValue: string | undefined) {
  if (dateValue == null) return;

  const parsedDate = parse(dateValue, "yyyyyy-MM-dd", new Date());

  // Check for Invalid Date
  if (isNaN(+parsedDate)) return;
  return parsedDate;
}
```

Reply:
Ok nice!

Reply:
We can actually do the same for 

```js
// As per https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#times
export function parseTime(timeValue: string | undefined) {
  if (timeValue == null || !isString(timeValue)) return;

  const timeRegex = timeValue.match(/(\d\d)(:(\d\d)){1,2}/i);
  if (timeRegex == null) return;

  const time = timeValue.split(":");
  const date = new Date();

  if (+time[0] < 0 || +time[0] > 23 || +time[1] < 0 || +time[1] > 59) return;

  date.setHours(+time[0]);
  date.setMinutes(+time[1]);
  date.setSeconds(0);

  if (time[2]) {
    if (+time[2] < 0 || +time[2] > 60) return;
    date.setSeconds(+time[2]);
  }

  return date;
}
```

Since we acceps both HH:mm and HH:mm:ss, we need to chekc the regex & split them and then do the parse & check for invalid date.

Above will be more equivalent to our code in the lines of code. So I believe we no need any refactoring here.